geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/Z3Iterator.scala

```diff
@@ -8,88 +8,76 @@
-  private var rowToZ: Array[Byte] => Long = null
+  private var keyXY: String = _
+  private var keyT: String = _
+  private var filter: Z3Filter = null
 
   private var topKey: Key = null
   private var topValue: Value = null
-  private val row = new Text()
 
   override def init(source: SortedKeyValueIterator[Key, Value],
                     options: java.util.Map[String, String],
                     env: IteratorEnvironment): Unit = {
     this.source = source
 
-    zOffset = options.get(ZOffsetKey).toInt
-    zLength = options.get(ZLengthKey).toInt
-
-    rowToWeek = getRowToWeek(zOffset)
-    rowToZ    = getRowToZ(zOffset, zLength)
-
-    keyXY = options.get(ZKeyXY)
-    keyT  = options.get(ZKeyT)
-
-    xyvals = keyXY.split(TermSeparator).map(_.split(RangeSeparator).map(_.toInt))
-
-    minWeek = Short.MinValue
-    maxWeek = Short.MinValue
-
-    val weeksAndTimes = keyT.split(WeekSeparator).filterNot(_.isEmpty).map { times =>
+    var minEpoch: Short = Short.MinValue
+    var maxEpoch: Short = Short.MinValue
+    keyT = options.get(ZKeyT)
+    val epochsAndTimes = keyT.split(EpochSeparator).filterNot(_.isEmpty).map { times =>
       val parts = times.split(TermSeparator)
-      val week = parts(0).toShort
-      // set min/max weeks - note: side effect in map
-      if (minWeek == Short.MinValue) {
-        minWeek = week
-        maxWeek = week
-      } else if (week < minWeek) {
-        minWeek = week
-      } else if (week > maxWeek) {
-        maxWeek = week
+      val epoch = parts(0).toShort
+      // set min/max epochs - note: side effect in map
+      if (minEpoch == Short.MinValue) {
+        minEpoch = epoch
+        maxEpoch = epoch
+      } else if (epoch < minEpoch) {
+        minEpoch = epoch
+      } else if (epoch > maxEpoch) {
+        maxEpoch = epoch
       }
-      (week, parts.drop(1).map(_.split(RangeSeparator).map(_.toInt)))
+      (epoch, parts.drop(1).map(_.split(RangeSeparator).map(_.toInt)))
     }
 
-    tvals = if (minWeek == Short.MinValue) Array.empty else Array.ofDim(maxWeek - minWeek + 1)
-    weeksAndTimes.foreach { case (w, times) => tvals(w - minWeek) = times }
+    val tvals: Array[Array[Array[Int]]] =
+      if (minEpoch == Short.MinValue) Array.empty else Array.ofDim(maxEpoch - minEpoch + 1)
+    epochsAndTimes.foreach { case (w, times) => tvals(w - minEpoch) = times }
+
+    keyXY = options.get(ZKeyXY)
+    val xyvals = keyXY.split(TermSeparator).map(_.split(RangeSeparator).map(_.toInt))
+    val zOffset = options.get(ZOffsetKey).toInt
+    val zLength = options.get(ZLengthKey).toInt
+    filter = new Z3Filter(xyvals, tvals, minEpoch, maxEpoch, zOffset, zLength)
   }
 
   override def next(): Unit = {
     source.next()
     findTop()
   }
 
+  private val text = new Text
   private def findTop(): Unit = {
     topKey = null
     topValue = null
     while (source.hasTop) {
-      if (inBounds(source.getTopKey)) {
+      val row = source.getTopKey.getRow(text)
+      val bytes = row.getBytes
```

Comment:
looks like you could call `getRowData` which would avoid copying the row array - although it does create a new `ArrayByteSequence` instead. Dunno if that would be better...